File: PracticasProgramacion/Practica_26/python/manejador_txt.py
"""
ManejadorTXT: escribe y lee archivos de texto plano con datos basicos, arreglos, AutoADT y PersonaADT.
Formato legible por humanos, similar al C++.
"""

from dataclasses import dataclass
from typing import List


@dataclass
class AutoADT:
    precio: float
    anio: int


@dataclass
class PersonaADT:
    nombre: str
    ap: str
    am: str
    genero: str
    edad: int


class ManejadorTXT:
    def escribir_datos_basicos(self, nombre_archivo: str) -> None:
        with open(nombre_archivo, "w") as f:
            f.write("int: 42\n")
            f.write("double: 3.14\n")
            f.write("char: a\n")
            f.write("bool: true\n")

    def escribir_arreglo(self, nombre_archivo: str) -> None:
        arr = [1, 2, 3, 4, 5]
        with open(nombre_archivo, "w") as f:
            for i, v in enumerate(arr):
                f.write(f"arr[{i}] = {v}\n")

    def escribir_autos(self, nombre_archivo: str) -> None:
        autos: List[AutoADT] = [AutoADT(15000, 2020), AutoADT(25000, 2023)]
        with open(nombre_archivo, "w") as f:
            for i, a in enumerate(autos):
                f.write(f"Auto {i}: precio={a.precio} anio={a.anio}\n")

    def escribir_personas(self, nombre_archivo: str) -> None:
        personas: List[PersonaADT] = [
            PersonaADT("Juan", "Garcia", "Lopez", "M", 25),
            PersonaADT("Maria", "Perez", "Sanchez", "F", 30),
        ]
        with open(nombre_archivo, "w") as f:
            for i, p in enumerate(personas):
                f.write(f"Persona {i}: {p.nombre} {p.ap} {p.am}, {p.genero}, {p.edad} años\n")

    def leer_y_mostrar(self, nombre_archivo: str) -> None:
        try:
            with open(nombre_archivo) as f:
                for linea in f:
                    print(linea.rstrip())
        except OSError:
            print(f"No se pudo abrir {nombre_archivo}")

File: PracticasProgramacion/Practica_26/cpp/ManejadorTXT.cpp
#include "ManejadorTXT.h"
#include <fstream>
#include <iostream>

void ManejadorTXT::escribirDatosBasicos(const std::string& nombreArchivo) const {
    std::ofstream f(nombreArchivo);
    if (!f) return;
    f << "int: 42\n";
    f << "double: 3.14\n";
    f << "char: a\n";
    f << "bool: true\n";
}

void ManejadorTXT::escribirArreglo(const std::string& nombreArchivo) const {
    std::ofstream f(nombreArchivo);
    if (!f) return;
    int arr[] = {1, 2, 3, 4, 5};
    for (size_t i = 0; i < 5; ++i)
        f << "arr[" << i << "] = " << arr[i] << "\n";
}

void ManejadorTXT::escribirAutos(const std::string& nombreArchivo) const {
    std::ofstream f(nombreArchivo);
    if (!f) return;
    AutoADT autos[] = {{15000, 2020}, {25000, 2023}};
    for (size_t i = 0; i < 2; ++i)
        f << "Auto " << i << ": precio=" << autos[i].precio << " anio=" << autos[i].anio << "\n";
}

void ManejadorTXT::escribirPersonas(const std::string& nombreArchivo) const {
    std::ofstream f(nombreArchivo);
    if (!f) return;
    PersonaADT personas[] = {{"Juan", "Garcia", "Lopez", "M", 25}, {"Maria", "Perez", "Sanchez", "F", 30}};
    for (size_t i = 0; i < 2; ++i)
        f << "Persona " << i << ": " << personas[i].nombre << " " << personas[i].ap << " "
          << personas[i].am << ", " << personas[i].genero << ", " << personas[i].edad << " años\n";
}

void ManejadorTXT::leerYMostrar(const std::string& nombreArchivo) const {
    std::ifstream f(nombreArchivo);
    if (!f) {
        std::cout << "No se pudo abrir " << nombreArchivo << "\n";
        return;
    }
    std::string linea;
    while (std::getline(f, linea))
        std::cout << linea << "\n";
}
